text/0030-shared-event-queue.md

```diff
@@ -0,0 +1,122 @@
+- Feature Name: Shared ThreadsafeFunction in EventQueue
+- Start Date: 2021-05-19
+- RFC PR: (leave this empty)
+- Neon Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Make `EventQueue` use a shared `ThreadsafeFunction` to leverage the
+[performance optimization][0] in recent Node.js versions.
+
+# Motivation
+[motivation]: #motivation
+
+`EventQueue` provides a way to invoke Rust closures on a JavaScript thread from
+another thread by using a `ThreadsafeFunction`. Recent Node.js versions have a
+[performance optimization][0] for the case where a single `ThreadsafeFunction`
+is invoked multiple times in a row. Thus using a single shared
+`ThreadsafeFunction` in `EventQueue`s would lead to significant performance
+improvements for Neon users.
+
+These improvements are perhaps most noticeable when Neon addon is used from the
+Electron, because scheduling new UV ticks in Electron are very costly.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Whenever `cx.queue()` is invoked - a new instance of `EventQueue` is created.
+With ithe proposed changes the instance would point to a shared
+`ThreadsafeFunction` internally. For the most scenarios there would be no
+difference between current and proposed implementation. Calling `queue.send()`
+will still invoke the closure on the JavaScript thread:
+```rust
+let queue = cx.queue();
+
+std::thread::spawn(move || {
+    queue.send(|| {
+        Ok(())
+    })
+});
+```
+
+Calling `queue.send()` several times in a row using different `cx.queue()` would
+be significantly faster:
+```rust
+loop {
+    cx.queue().send(|| {
+        Ok(())
+    })
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+With the recent [node.js changes][0] `ThreadsafeFunction` will loop over pending
+calls, invoking up to 1000 of them in the same tick instead of running each new
+call on a separate UV loop tick. This is an implementation detail that is
+intentionally not covered by `napi_create_threadsafe_function` documentation to
+allow for potential performance optimizations in the future.
+
+Neon doesn't currently leverage it without the proposed changes because a new
+`ThreadsafeFunction` is created for each new `EventQueue`, and `cx.queue()`
+creates a new `EventQueue` every time as requested. So instead of doing that it
+is proposed that we use existing `lifetime::InstanceData` internal API layer to
+hold a wrapper around `ThreadsafeFunction` tailored to execute Rust closures
+specifically, and to use an `Arc` reference to a `RwLock` over this wrapper in
```

Comment:
Is a `RwLock` necessary? The only non-threadsafe operations on `EventQueue` are `unref` and `reference`. Both of these require a `Context` which already enforces single threaded-ness.

Reply:
The `.reference()`/`.unref()` methods of `ThreadsafeFunction` have `mut self`, so `ThreadsafeTrampoline` needs to be mutable in order to call them. Is there any better way of doing it?